### ISSUE TITLE
Hook after renderToString and/or app.mount

### ISSUE THREAD

**Issue by wucdbm** [frandiox/vite-ssr#84]
Heya, me again.

(vue.js 3.0 app)

I was reading the source code and noticed that our initialization hook is run early on, which is great, however, in my application (similar to https://pinia.esm.dev/ssr/#state-hydration example) I have my custom data stores.
Trouble is, those are filled with data by components at arbitrary levels deep, and I only have the data once the app has been fully rendered, as I rely on `onServerPrefetch(async () => {})`

That means that the `initialState` variable passed to the userland hook can't have the final data the server has fetched, as my hook is ran very early on. I need another hook right after this await

```
renderToString(app, context).then(deferred.resolve).catch(deferred.reject)
const body = await deferred.promise
```

To make things consistent on the client end, we could be running that hook after `app.mount('#app', true)`

Other than that, using this library has been a great experience so far!

**Comment by frandiox** [comment]
@wucdbm Hey thanks. What's your use case for this exactly?
There's a `transformState` hook that runs at the very end and gets the filled `initialState` as its first argument. This is normally used to make arbitrary modifications to the initial state and it can also be asynchronous. Therefore, I think this can be used in most situations.

Would that work for you?

**Comment by wucdbm** [comment]
I'll have a look next monday, that could very well be something that I overlooked!

**Comment by wucdbm** [comment]
@frandiox Priorities shifted and it was left behind, I'll get back to the thread as soon as I'm back to that project
It's literally the only thing left before we have a stable project bootstrap and we must evaluate it against a custom solution

**Comment by wucdbm** [comment]
@frandiox No, unfortunately, that hook will **not** help, let me attemp to describe the issue

My main.ts looks like this

```
export default viteSSR(
    App,
    options,
    ({ app, router, initialState, axiosBaseUrl, axiosHost, request }: HookParams) => {"
    // some initialization code
    // ...
    
        const ssrCache = ssrExchange({
            isClient: !import.meta.env.SSR,
            // eslint-disable-next-line @typescript-eslint/ban-ts-comment
            // @ts-ignore
            initialState: !import.meta.env.SSR ? window.__INITIAL_STATE__?._urql : undefined,
            staleWhileRevalidate: true,
        })
        
        app.use(urql, {
            url: 'https://rickandmortyapi.com/graphql',
            exchanges: [
                dedupExchange,
                cacheExchange,
                ssrCache,
                fetchExchange,
            ],
        })
```

The issue here is, that in my components, I do this

```
<script lang="ts">
import { defineComponent, onServerPrefetch } from 'vue'
import { getCharacters } from '@/gql/queries'
import { useQuery } from '@urql/vue'

export default defineComponent({
    name: 'HelloWorldComp',
    setup() {
        const result = useQuery({ query: getCharacters })

        onServerPrefetch(async () => {
            await result
        })

        return {
            fetching: result.fetching,
            data: result.data,
            error: result.error,
        }
    },
})
</script>
```

The thing is, I'll have the data in the `ssrCache` I passed to urql earlier ONLY AFTER I have awaited all of its promises via `onServerPrefetch`.
So in that sense, I need to be able to, in addition to `head`, return a hook that will be ran by entry-server just after `renderToString(app, context).then(deferred.resolve).catch(deferred.reject)`

Then I am guaranteed that all of my promises will have been awaited, and then I still have access and can modify `initialState`, like so: 
```
    if (ssrCache) {
        initialState._urql = ssrCache.extractData()
    }
```

(The reason I am not replacing initialState entirely is because I have other custom stores that hold other bits of data, these are simply fields in initialData and allows for future flexibility)

Essentially I am looking forward to a change such as 

```
const { head, postRenderHook } =
      (hook &&
```

```
renderToString(app, context).then(deferred.resolve).catch(deferred.reject)
const body = await deferred.promise
(postRenderHook && (await postRenderHook()) // << this should do the magic trick if the user of the library has exported this postRenderHook from their main.ts hook
```

**Comment by frandiox** [comment]
@wucdbm  Yeah I think I understand the use case. I've done this with Apollo before so I think it should work the same way with URQL. 

Can you test something like this?

```js
export default viteSSR(
  App,
  {
    routes,
    transformState(state, defaultTransformer) {
      if (import.meta.env.SSR) {
        // Make cache serializable in SSR
        state.urqlCache = state.urqlCache.extractData()
      }

      // Continue default serialization/deserializaion
      return defaultTransformer(state)
    },
  },
  ({ app, initialState }) => {
    // Main initialization hook
    
    initialState.urqlCache = ssrExchange({
        isClient: !import.meta.env.SSR,
        initialState: !import.meta.env.SSR ? initialState.urqlCache : undefined,
        staleWhileRevalidate: true,
    })
    
    app.use(urql, {
        url: 'https://rickandmortyapi.com/graphql',
        exchanges: [
            dedupExchange,
            cacheExchange,
            initialState.urqlCache,
            fetchExchange,
        ],
    })
  }
)
```


The key here is to save the ssrCache in the initial state object so you have access to it later in the `transformState` to serialize it.

There are other examples that follow this strategy with [apollo](https://vitedge.js.org/recipes/react.html#apollo-graphql) and [vue-query](https://vue-query.vercel.app/#/guides/ssr?id=using-vite-ssr).

Please tell me if something like that works for you.

**Comment by wucdbm** [comment]
@frandiox That'd definitely work, yes, but looks ugly -- IMO we shouldn't mess with business objects attached to a container that contains pure state only
On one hand, with my solution it'd need some more documentation or people to browse the source code to figure things out (I've done a lot of that on this and other libraries), on the other though, as a project evolves and becomes more complex, this hacking around may end up a huge shotgun wound in your foot

It's your call ultimately, but I'd keep the state object and other containers that implement business logic completely separate

**Comment by wucdbm** [comment]
@frandiox Just to confirm, your solution works fine

**Comment by frandiox** [comment]
The problem in returning that hook is that it's not isometric because it would not be used in the browser.
It would be nice if the cache instance implemented a `.toJSON` method that is automatically called by `JSON.stringify` 🤔 

Regardless, I'll give it a thought anyway since it's always possible to improve the API to something that is easier to understand.

Closing this for now since the main issues is solved 👍

**Comment by frandiox** [comment]
@wucdbm One more thing, can you try something like this?

```js
  ({ app, initialState, isClient }) => {
    // Main initialization hook
    
    const ssrCache = ssrExchange({
        isClient,
        staleWhileRevalidate: true,
        initialState: initialState._urql, // undefined in SSR
    })

    if (import.meta.env.SSR) {
        ssrCache.toJSON = () => ssrCache.extractData()  // or ssrCache.toJSON = ssrCache.extractData , not sure
        initialState._urql = ssrCache
    }
    
    app.use(urql, {
        url: 'https://rickandmortyapi.com/graphql',
        exchanges: [
            dedupExchange,
            cacheExchange,
            ssrCache,
            fetchExchange,
        ],
    })
  }
```

Not sure if this is easier but at least it doesn't need to define `transformState` 🤔

**Comment by wucdbm** [comment]
Yup, the one in the comment with `toJSON` is actually great

**Comment by frandiox** [comment]
@wucdbm Do you think that would be a better fit for official examples? 🤔 (instead of using transformState hook)

**Comment by wucdbm** [comment]
I think both -- some people may not like the "black magic" - If anything, at least include a link to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#description
I wasn't aware of it myself, actually, but guessed it's something like that based off of your example and found the documentation later

**Comment by frandiox** [comment]
@Moeriki  Ah right, this is only for JSON states. I think devalue outputs a JS object or function instead.
Out of curiosity, why do you need devalue? Are you serializing Map/Set, regexps or something complex?
